geolocator_android/android/src/main/java/com/baseflow/geolocator/permission/PermissionManager.java

```diff
@@ -124,67 +132,89 @@ public boolean onRequestPermissionsResult(
       return false;
     }
 
-    LocationPermission permission = LocationPermission.denied;
-
-    int requestedPermissionIndex = indexOf(permissions, requestedPermission);
-
-    if (requestedPermissionIndex < 0) {
-      Log.w(
-          "Geolocator",
-          "Location permissions not part of permissions send to onRequestPermissionsResult method.");
-      return false;
-    }
-
     if (grantResults.length == 0) {
       Log.i(
           "Geolocator",
           "The grantResults array is empty. This can happen when the user cancels the permission request");
       return false;
     }
 
-    if (grantResults[requestedPermissionIndex] == PackageManager.PERMISSION_GRANTED) {
-      if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
-        int backgroundPermissionIndex =
-            indexOf(permissions, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
-        if (backgroundPermissionIndex >= 0
-            && grantResults[backgroundPermissionIndex] == PackageManager.PERMISSION_GRANTED) {
-          permission = LocationPermission.always;
-        } else {
-          permission = LocationPermission.whileInUse;
-        }
-      } else {
-        permission = LocationPermission.always;
+    LocationPermission locationPermission = LocationPermission.denied;
+    int grantedResult = PackageManager.PERMISSION_DENIED;
+    boolean shouldShowRationale = false;
+    boolean permissionsPartOfPermissionsResult = false;
+
+    for (String permission : requestedPermissions) {
+      int requestedPermissionIndex = indexOf(permissions, permission);
+      if (requestedPermissionIndex > 0) {
+          permissionsPartOfPermissionsResult = true;
+      }
+      if (requestedPermissionIndex == PackageManager.PERMISSION_GRANTED) {
+        grantedResult = PackageManager.PERMISSION_GRANTED;
       }
+      if (ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) {
+        shouldShowRationale = true;
+      }
+    }
+
+    if(!permissionsPartOfPermissionsResult) {
+        Log.w(
+                "Geolocator",
+                "Location permissions not part of permissions send to onRequestPermissionsResult method.");
+        return false;
+    }
+
+    if (grantedResult == PackageManager.PERMISSION_GRANTED) {
+        locationPermission = (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.Q || hasBackgroundAccess(permissions, grantResults))
+                ? LocationPermission.always
+                : LocationPermission.whileInUse;
     } else {
-      if (activity != null && !ActivityCompat.shouldShowRequestPermissionRationale(activity, requestedPermission)) {
-        permission = LocationPermission.deniedForever;
+      if (activity != null && !shouldShowRationale) {
```

Comment:
Checking if `activity` is `null` or not is not relevant here as it is not used in the context of this `if` statement or its body:

```suggestion
      if (!shouldShowRationale) {
```